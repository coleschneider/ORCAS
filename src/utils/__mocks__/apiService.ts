export default {
  post: (arg: any) =>
    Promise.resolve({
      data: {
        message: 'test',
      },
    }),
};
