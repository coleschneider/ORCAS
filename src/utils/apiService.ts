import axios from 'axios';

export const instance = axios.create({
  baseURL: process.env.LAMBDA_CDN
})