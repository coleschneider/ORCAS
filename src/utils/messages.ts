import { toast } from 'react-toastify';
import * as _ from 'lodash';
export const NETWORK_MESSAGES = {
  offline: {
    text:
      "Experiencing connection problems? The app will still work offline, but you won't be viewing the most recent version",
    style: 'warn',
  },
  online: {
    text: "You're back online",
    style: 'success',
  },
};

export const NetworkNotification = (type, opts = {}) =>
  toast(
    NETWORK_MESSAGES[type],
    _.defaults(
      {
        // className: 'network-msg container',
        // bodyClassName: 'network-msg content',
        // progressClassName: 'network-msg progress'
      },
      opts,
    ),
  );
