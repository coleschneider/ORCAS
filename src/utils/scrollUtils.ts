export const createScroller = (
  observerCb: (el: Element) => (IntersectionObserverCallback) => void,
  options: IntersectionObserverInit,
) =>
  document.querySelectorAll('.target-section').forEach(el => {
    const checkMission = observerCb;
    const cb = checkMission(el);
    const observer = new IntersectionObserver(cb, options);
    return observer.observe(el);
  });
