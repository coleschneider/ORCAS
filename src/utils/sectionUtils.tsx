import * as React from 'react';
interface SectionProps {
  sectionName: string;
  description?: string;
  displayTitle: string;
  className?: string;
}
export const createSection = ({ sectionName, className, description, displayTitle, ...rest }: SectionProps) => ({
  children,
}) => (
  <section id={sectionName} className={`s-${className ? className : sectionName} target-section`}>
    <div className="row section-header">
      <div className="col-full">
        <h1 className="display-1">{displayTitle}</h1>
        {description && <p className="lead">{description}</p>}
      </div>
    </div>
    {typeof children === 'function' ? children(rest) : children}
  </section>
);

export const ServiceItem = ({ icon, title, children }) => (
  <div className="col-block item-feature">
    <div className="item-feature__icon">
      <i className={icon} />
    </div>
    <div className="item-feature__text">
      <h3 className="item-title">{title}</h3>
      <p>{children}</p>
    </div>
  </div>
);
