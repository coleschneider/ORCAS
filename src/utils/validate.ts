import { FormikErrors } from 'formik';
interface FormValues {
  amount?: string;
}
export function validateDonation(values: FormValues) {
  const errors: FormikErrors<FormValues> = {};
  if (!values.amount) {
    errors.amount = 'Required';
  } else if (!/^[0-9]+(\.[0-9]{1,2})?$/.test(values.amount)) {
    errors.amount = 'Please enter a valid US Dollar amount';
  }
  return errors;
}
