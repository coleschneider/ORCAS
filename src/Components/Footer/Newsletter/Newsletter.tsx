import * as React from 'react';
import { Formik } from 'formik';
import { ToastContainer, toast } from 'react-toastify';
import './newsletter.scss';
import { validateSubscribe } from 'utils/validate';
import instance from 'utils/apiService';

interface FormFields {
  email_address?: string;
}
interface NewsletterState {
  isFetching: boolean;
  error: null | Error;
}
class NewsletterForm extends React.Component<{}, NewsletterState> {
  constructor(props) {
    super(props);
    this.state = {
      error: null,
      isFetching: false,
    };
  }
  notifySuccess = ({ message }) =>
    toast.success(message, {
      autoClose: 5000,
      className: 'notification-formSuccess',
      position: 'bottom-right',
    });
  notifyError = () =>
    toast.warn(`There was an error attempting to subscribe to newsletter`, {
      autoClose: 5000,
      className: 'notification-formError',
      position: 'bottom-right',
    });
  subscribeUser = async (values: FormFields, cb) => {
    this.setState({
      isFetching: true,
    });
    return instance
      .post('/', values)
      .then(res => {
        this.setState({
          isFetching: false,
        });
        this.notifySuccess(res.data);
        cb();
      })
      .catch(error => {
        this.notifyError();
        this.setState({
          error,
          isFetching: false,
        });
      });
  };
  render() {
    const { isFetching } = this.state;
    return (
      <div className="subscribe-form">
        <Formik initialValues={{ email_address: '' }} onSubmit={this.subscribeUser} validate={validateSubscribe}>
          {({
            values,
            touched,
            handleSubmit,
            errors,
            isSubmitting,
            handleChange,
            handleBlur,
            resetForm,
            initialValues,
          }) => {
            return (
              <form
                onSubmit={e => {
                  handleSubmit(e, resetForm);
                }}
                id="mc-form"
                className="group"
                test-id="contact-form"
              >
                <input
                  type="email"
                  name="email_address"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.email_address}
                  autoComplete="off"
                  className="email_address"
                  id="mc-email"
                  placeholder="Email Address"
                  className={errors.email_address && touched.email_address ? 'email text-input error' : 'email'}
                />
                {errors.email_address && touched.email_address && (
                  <div test-id="contact-form-errors" className="input-feedback">
                    {errors.email_address}
                  </div>
                )}

                <button test-id="newsletter-btn" className="btn" type="submit" disabled={isSubmitting || isFetching}>
                  Submit
                </button>
                <label className="subscribe-message" />
              </form>
            );
          }}
        </Formik>
        <ToastContainer />
      </div>
    );
  }
}

export default NewsletterForm;
