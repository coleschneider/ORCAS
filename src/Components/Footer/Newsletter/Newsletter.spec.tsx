import * as React from 'react';
import Newsletter from './Newsletter';
import apiService from 'utils/apiService';
import * as sinon from 'sinon';
import * as validation from 'utils/validate';

jest.mock('utils/validate');

jest.mock('utils/apiService');

describe('Newsletter', () => {
  let wrapper;
  const submitForm = async (email = 'testing@gmail.com') => {
    wrapper = mount(<Newsletter />);
    await wrapper
      .find('Formik')
      .props()
      .onSubmit(email);
  };
  it('should match the snapshot', () => {
    const wrapper = shallow(<Newsletter />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Newsletter Form', () => {
    it('should call subscribeUser', async () => {
      const subscribeSpy = jest.spyOn(apiService, 'post');
      const wrapper = mount(<Newsletter />);
      await wrapper
        .find('Formik')
        .props()
        .onSubmit('testing@gmail.com');

      expect(subscribeSpy).toHaveBeenCalled();
      expect(subscribeSpy).toHaveBeenCalledWith('/', 'testing@gmail.com');
    });

    it('should call validate when the form changes', async () => {
      const validateSpy = jest.spyOn(validation, 'validateSubscribe');
      const wrapper = mount(<Newsletter />);
      wrapper.find('input').simulate('change', {
        target: {
          name: 'email_address',
          value: 'testing@gmail.com',
        },
      });
      wrapper
        .find('Formik')
        .props()
        .validate({ email_address: 'asdsad' });
      expect(validateSpy).toHaveBeenCalled();
    });
  });
  describe('Form Messages', () => {
    const clock = sinon.useFakeTimers();
    let apiInstanceSpy;

    it('should handle succes messages correctly', async () => {
      apiInstanceSpy = jest.spyOn(apiService, 'post');
      await submitForm();
      clock.tick(5);
      wrapper.update();
      expect(apiInstanceSpy).toBeCalledWith('/', 'testing@gmail.com');
      expect(
        wrapper
          .find('.Toastify__toast-body')
          .first()
          .text(),
      ).toBe('Successfully subscribed to newsletter!');
    });
    it('should handle error messages correctly', async () => {
      apiInstanceSpy = jest.spyOn(apiService, 'post').mockImplementation(() => Promise.reject());
      await submitForm();
      clock.tick(5);
      wrapper.update();
      expect(wrapper.find('.Toastify__toast-body').text()).toBe(
        'There was an error attempting to subscribe to newsletter',
      );
    });
  });
});
