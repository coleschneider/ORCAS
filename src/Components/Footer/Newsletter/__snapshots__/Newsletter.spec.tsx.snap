// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter should match the snapshot 1`] = `
<div
  className="subscribe-form"
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "email_address": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    validate={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <Component />
  </Formik>
  <ToastContainer
    autoClose={5000}
    bodyClassName={null}
    className={null}
    closeButton={
      <CloseButton
        ariaLabel="close"
      />
    }
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName={null}
    progressStyle={null}
    rtl={false}
    style={null}
    toastClassName={null}
    transition={[Function]}
  />
</div>
`;
